### DISCUSSION TITLE
Rodauth usage with React Native Applications

### DISCUSSION THREAD

**Discussion by doloboyz** [jeremyevans/rodauth#327] [General]
Hello, 
I was considering using rodauth as an authentication solution for a startup and was looking for info regarding implementation in a react native ios application. As far as I've seen, rodauth docs mention use in web apps; however, I've seen no mention of use in native apps.

Is rodauth suitable for use in react native applications? If so, how would I go about dealing with react native views rather than html? Is rodauth suitable for use in a business environment or does this library target personal projects? What would auth flow look like for a native application using rodauth?

I think I had a few more questions, just can't remember them right now, so I'll comment them once they come back to me. 

Thanks for any help or advice!

**Comment by jeremyevans** [comment]
Yes, you can use Rodauth as a backend for native applications.  You'll want to use the json feature, and potentially the jwt feature (if you want to use JWTs for storage instead of session cookies).  This allows for a full JSON API (no HTML).  You would be responsible for handling the react native views yourself in this case.

Rodauth is suitable to use both in a business environment and personal projects.  It is very customizable, and any part can generally be easily overridden as needed.

In terms of auth flow, it heavily depends on what type of auth you are doing.  For simple login/password, there would be a POST request to test whether the login/password is valid.  Rodauth is very flexible and there are many additional types of auth it supports (multifactor auth, passwordless auth with email links or webauth).  You can see the specs in the Rodauth repository for how to use Rodauth's JSON API for each Rodauth feature (all Rodauth features supported via normal browser requests are supported via a JSON API if the json feature is enabled).

I recommend you review Rodauth's website (https://rodauth.jeremyevans.net), and you can post here if you have additional questions that the documentation doesn't cover.

**Reply by doloboyz** [reply]
Thank you so much for the quick response! I'll be sure to spend the next few days reading through the docs. 

I also was wondering how you feel about the ecosystem built around rodauth (specifically rodauth-oauth and rodauth-rails). Do you think these extensions are mature enough to use in a startup as well? 

Also, I know this may be a tough question to answer considering you maintain rodauth and may have some bias, however, I was wondering what the drawbacks of using rodauth are? Do you feel there's any areas in which rodauth is lacking or that there are any serious cons associated with using rodauth as opposed to another 3rd-party solution or custom solution? 

Thanks again for your time!

**Reply by jeremyevans** [reply]
I don't have a need for OAuth or Rails in my apps, so I haven't personally used rodauth-oauth or rodauth-rails.  If I had a need for either OAuth or Rails, I would probably use them.

Drawbacks of using Rodauth:

* Additional dependencies if running Rails (Roda/Sequel)
* No built-in support for OAuth, SAML, and potentially other things
* Internally implemented using a DSL that requires somewhat of a learning curve

On the flip site:

* Both Roda and Sequel are fairly lightweight compared to alternatives such as Rails and ActiveRecord
* What is supported works well and has 100% line/branch coverage (I don't accept features I'm not comfortable maintaining)
* Internal DSL significantly DRYs up code, resulting in easier maintenance once it is understood

In terms of Rodauth compared to a 3rd-party solution, it mostly depends on how comfortable you are running things yourself, and also depends on if the 3rd-party solution has features Rodauth lacks that you plan to use.

In terms of Rodauth compared to a custom solution, it would be a lot of work to implement all of Rodauth's features.  However, if your needs are fairly simple (just login/password auth), a custom developed solution isn't terribly difficult.  However, if you've never developed authentication code before, you may not be aware of all the security pitfalls involved.
